sim/simx/execute.cpp

```diff
@@ -1414,6 +1415,177 @@ void Emulator::execute(const Instr &instr, uint32_t wid, instr_trace_t *trace) {
+
+          //Store C
+          for (int n=0; n<num_data_per_thread_st; n++)
+          {
+            uint64_t mem_addr = (base_addr+(n*mem_bytes));
+            uint32_t csr_index = (2*num_data_per_thread_st) + n;
+            uint32_t scratchpad_index = (tc_size*tc_size*2) + (t*num_data_per_thread) + n;
+            
+            //scratchpad -> csr (TODO :: removed intermediate CSR stage ; incorporate limited scratchmad implementation)
+            //core_->set_csr(csr_addr[(2*num_data_per_thread) + n], scratchpad[(n_tiles*tc_size*tc_size*2) + (t*num_data_per_thread) + n], t, warp_id_);
+            Word* temp_ref = &(warp.ireg_file.at(t).at(rsrc0));
+            *temp_ref = scratchpad[(n_tiles*tc_size*tc_size*2) + (t*num_data_per_thread_st) + n];
+
+            this->dcache_write(temp_ref, base_addr+(n*mem_bytes), mem_bytes);  
+          }
+        }
+        //Clear the scratchpad
+        for(int i =0 ; i < scratchpad.size(); i++)
+        {
+          scratchpad[i] = 0;
+        }
+      }
+      break;
+      case 2: 
+      { //Matrix Multiply
+        DP(4, "TCU MULTIPLY MAT");
+        trace->fu_type = FUType::TCU;
+        trace->tcu_type = TCUType::TCU_MUL;
+        uint32_t accu_offset = (n_tiles)*(n_tiles)*(n_tiles)*tc_size*tc_size*2;
+        uint32_t threads_per_tc = MAX (1, num_threads/TC_per_warp);
+        for (uint32_t t = thread_start; t < num_threads_actv; ++t) 
+        {
+          if (!warp.tmask.test(t))
+            continue;
+         
+          DP(3, "Thread ID" << t); 
+          //TC operation [only 1 thread in 1 warp needs to do this]
+          if (t%threads_per_tc == 0)
+          {
+            //TODO : change to systolic array implementation
+            uint32_t thread_offset = t*(tc_size*tc_size);
+            int loop_offset = 0;
+            int offset_b = n_tiles*n_tiles*n_tiles*tc_size*tc_size;
+            /*
+            // TODO : Fix needed for functional correctness
+            for(int tiles = 0 ; tiles < n_tiles ; tiles++)  //What's the HW implication of this?? A counter implementation?
+            { 
+              for (int i = 0; i < tc_size; i++) { //ROW-1
+                for (int j = 0; j < tc_size; j++) { //COL-2
+                  int sum = 0;
+                  for (int k = 0; k < tc_size; k++)
+                  { //COL-1
+                    sum = sum + scratchpad[loop_offset + thread_offset*n_tiles + i * tc_size + k] *scratchpad[loop_offset + thread_offset*n_tiles + offset_b + (k * tc_size + j)];
+                  }
+                  scratchpad[accu_offset + thread_offset +(i * tc_size + j)] += sum; //[i * col2 + j] = sum
+                  DP(3, "Scratchpad Index: " << accu_offset + (i * tc_size + j) << " , Value=" << scratchpad[accu_offset + (i * tc_size + j)]);
+                }
+              }
+              loop_offset += tc_size*tc_size; //Move to the next tiled matmul fragment
+            }
+            */
```

Comment:
So there is no matrix multiplication implemented?

Reply:
Explained in the last comment.